\name{plot.filtered}
\alias{plot.filtered}
\title{Diagnostic plot for filtered genes}
\usage{
  plot.filtered(x, y, output = "x11", path = NULL, ...)
}
\arguments{
  \item{x}{an annotation data frame like the ones produced
  by \code{\link{get.annotation}}. x should be the filtered
  annotation according to metaseqr's filters.}

  \item{y}{an annotation data frame like the ones produced
  by \code{\link{get.annotation}}. x should contain the
  total annotation without the application of any metaseqr
  filter.}

  \item{output}{one or more R plotting device to direct the
  plot result to. Supported mechanisms: "x11" (default),
  "png", "jpg", "bmp", "pdf" or "ps".}

  \item{path}{the path to create output files.}

  \item{...}{further arguments to be passed to plot
  devices, such as parameter from \code{\link{par}}.}
}
\value{
  The filenames of the plots produced in a named list with
  names the which.plot argument. If output="x11", no output
  filenames are produced.
}
\description{
  This function plots a grid of four graphs depicting: in
  the first row, the numbers of filtered genes per
  chromosome in the first column and per biotype in the
  second column. In the second row, the percentages of
  filtered genes per chromosome related to the whole genome
  in the first columns and per biotype in the second
  column.
}
\examples{
\dontrun{
y <- get.annotation("mm9","gene")
x <- y[-sample(1:nrow(y),10000),]
plot.filtered(x,y)
}
}
\author{
  Panagiotis Moulos
}

